### PR TITLE
Opaque occlusion culling

### LINKED ISSUE #1051: Optimization: Do not render items covered by opaque items
We often have screens where pages goes on top of eachother, and we want to not render these items at all.
We could have a opaque detection at several level:
 1. In the backend while doing the rendeing. femtovg can use a z buffer. in the software renderer, we can process the spans front to back first too to detect which regions are covered.
 2. we can also do a first pass before the processing.

The algorithm for 2 would be the following:

## Mark opaque regions

 - Visit items front to back. When we see an opaque item, use a space partition data structure to mark opaque region belonging to that item. (this can be done without registering dependencies)
 - When doing the rendering 
     * if a region is marked as opaque belonging to another item: don't do the rendering
     * if a region is marked as opaque belonging to the item to render: render the item and mark the space as not opaque
     * otherwise simply render the item

## Filter sibiling items

This would rely on the compiler to detect items for which the children are in bounds (most items) and also items whose sibling might overlap.

When rendering an item whose children may overlap:
 - instead of simply recursing into it, loop over the direct children in reverse order (front to back). If one of them is opaque and cover the whole item, then break the loop and start rendering from there only.

This should detect the common case of "StackView" kind of thing, but would miss a lot of other cases.  But does not need to keep a mapping between the region and the most opaque item.

### PR DESCRIPTION
**Pull request by R-Cramer4**
Closes #1051

`PartialRenderer::compute_occlusion` does a postorder walk over the same reverse-paint-order tree already used for painting, accumulating the screen regions covered by opaque, axis-aligned rectangles (`is_opaque_covering_rectangle`). Once an item falls fully inside the accumulated region, it's marked `occluded`, and `filter_item` skips drawing it.

### My design decisions:
- I added an `occlusion-culling` feature due to the fact that this check adds more work which could be not needed for scenes with no occlusion
- Only opaque rectangles contribute to occlusion, never a container's own bounding box. An unclipped child can paint outside its parent's rect, so treating "this container's bounds are fully covered" as "everything under it is hidden" would wrongly cull unclipped, overflowing children. Covered here by `unclipped_overflowing_children_are_not_culled`.
- Rounded clip: true containers inset their occluded region by the corner radius. An opaque, unrounded child that exactly fills a rounded clip's rectangular bounds still leaves the rounded-off corners unpainted (the clip masks to the rounded shape at render time). Without the inset, content behind those corners would be wrongly culled. Covered by `rounded_clip_does_not_cull_content_behind_its_corners`.
- Occlusion is region-based, not sibling-only. Occluder and occluded content can live in entirely different subtrees (e.g. a full-screen overlay sibling to an ancestor). Accumulating regions over the whole walk, rather than only comparing same-parent siblings, is what makes that case work. Covered by `opaque_cover_culls_hidden_content_in_a_different_branch`.

### Performance testing
I tested a scene with a 40x50 grid in a 1200x1200 window that rerenders every items color and text every frame, with being able to hide all of them behind an opaque item. This was tested with the `occlusion-culling` feature on and off along with the opaque item present and not present. This was ran 20 times in separate cargo bench processes, pinned to one CPU core to make noise less visible. This was tested based on master at commit `141a6ce`.

#### Results
| scene | config | mean | median | min | max | stdev | change vs. off (mean) |
|---|---|---|---|---|---|---|---|
| occluded_dashboard | off | 51.711 ms | 51.031 ms | 49.873 ms | 60.419 ms | 2.350 ms | -- |
| occluded_dashboard | on | 41.909 ms | 41.941 ms | 40.630 ms | 45.309 ms | 1.031 ms | -19.0% |
| unoccluded_dashboard | off | 55.832 ms | 55.200 ms | 53.633 ms | 63.909 ms | 2.473 ms | -- |
| unoccluded_dashboard | on | 56.057 ms | 55.313 ms | 53.854 ms | 60.223 ms | 2.011 ms | +0.4% |

### CODE REVIEW

**Comment by dfaure-kdab** [comment]
Claude Opus 5's review of PR #13027 "Opaque occlusion culling", at head `ae4de8fb7f`.

The algorithm itself is sound; the integration with `PartialRenderingState`'s frame lifecycle is what isn't.

## Critical

**1. Stale `occluded` flags erase content.** `compute_occlusion` runs inside the component loop at `internal/core/partial_renderer.rs:1185`, *before* `apply_dirty_region` does `force_screen_refresh.take()` (line 1191) and unions in `dirty_region_of_existing_buffer`. `compute_occlusion_recursive` prunes subtrees that miss the *pre-union* region, leaving `entry.occluded` at last frame's value; `filter_item` (~line 1060) then tests it against the *post-union* region.

Failure scenario: an `if`/repeater destroys an opaque full-screen overlay. `free_graphics_resources` frees only the destroyed entries and sets `force_screen_refresh` — nothing dirties the overlay's old rect. The next frame the dirty region is empty, so every surviving item is pruned, the content behind the overlay keeps `occluded = true` from the previous frame, and the full-screen repaint draws nothing there. The content stays missing until something else dirties it.

On Skia this is the normal path, not an exotic one: `render_components_to_canvas` passes the whole window as `buffer_dirty_region` whenever `back_buffer_age == 0`, so occlusion is always computed against a different region than the one it is consumed against. Steady-state frames already run on stale flags and only look right because nothing changed.

Suggested fix: move the `compute_occlusion` loop to the end of `apply_dirty_region`, after `force_screen_refresh.take()` and after the `dirty_region_of_existing_buffer` union. Any subtree still pruned is then by construction outside the region that will be drawn, and `filter_item`'s own dirty check makes a stale flag inert. Please add a test that removes an opaque overlay via an `if` and asserts the content behind it is drawn.

**2. The feature and its four tests never build or run in CI.** `grep -r occlusion .github/workflows` returns nothing. No job enables `i-slint-core/occlusion-culling`, and both test files start with `#![cfg(all(feature = "occlusion-culling", feature = "testing"))]`, so they compile to an empty binary in every existing job — the feature is not even compile-checked. Please add a job (or extend a feature-matrix job) running `cargo test -p i-slint-renderer-software --features testing,occlusion-culling` and at minimum `cargo check -p i-slint-core --features occlusion-culling`.

## Major

**3. Occluder rectangles are recorded in logical coordinates with no device-pixel rounding.** `partial_renderer.rs:568` records `occluder_rect` verbatim, and `contains_box` is an exact float containment test on logical coords — but the renderer rasterizes at `scale_factor` with anti-aliasing. An occluder whose logical edge lands on a fractional device pixel covers that boundary pixel at partial alpha; the occludee sharing that pixel is culled anyway, so the pixel blends the occluder over the window background instead of over the occludee. Visible 1px seams at fractional scale factors (`SLINT_SCALE_FACTOR=1.3`) and at any non-integral logical geometry. Shrink to the device-pixel interior before recording — `((rect * scale_factor).to_box2d().round_in() / scale_factor)` — and leave `contains_rect`'s test on the occludee un-shrunk. The `Layer` (`cache-rendering-hint`) path compounds this: `render_and_blend_layer` blits the cached image with `FilterMode::Linear`, so an opaque rect inside a `Layer` doesn't cover its logical rect exactly either.

**4. All four tests are too loose to fail on the regressions they document.** `unclipped_overflowing_children_are_not_culled` asserts `count >= OVERFLOW_RECT_COUNT` (20) on a scene with ~22 counted items, so wrongly culling two probes still passes. Same for `rounded_clip_does_not_cull_content_behind_its_corners` (`>= 25` with ~27 items) and, in the other direction, `opaque_cover_culls_hidden_siblings` (`< 50` passes with 49 of 50 rects still drawn). The scenes are fully deterministic, so please assert exact counts with `assert_eq!`.

**5. `slint-interpreter` as a dev-dependency of two leaf renderer crates.** `internal/renderers/skia/Cargo.toml` and `internal/renderers/software/Cargo.toml` now dev-depend on `slint-interpreter`, which depends on `i-slint-backend-selector`, which depends back on both renderers. Cargo tolerates the dev-dep cycle, but `cargo test -p i-slint-renderer-software` now builds the entire compiler, interpreter and winit backend — for a crate that is otherwise no_std/MCU-targetable. These are integration tests over the *core* partial renderer, not over Skia or the software rasterizer specifically; they belong in the `tests/` workspace, where the interpreter is already a dependency.

**6. The two test files are near-verbatim copies** (280 vs 267 lines): identical `.slint` sources, doc comments, assertion messages and constants, differing only in the platform harness. Extract the scenes and assertions into a shared module and parameterize on the harness — or, per #5, write it once in `tests/`.

**7. `compute_dirty_regions` is rewritten wholesale, and the rewrite predates master.** `compute_dirty_regions_recursive` (`partial_renderer.rs:687`) replaces the `visit_items` combinator with a hand-rolled `visit_children_item` recursion, restructuring ~200 lines of subtle, hot code (mostly re-indentation) purely so `subtree_screen_bounds` can be cached bottom-up. That bound only serves the pruning optimization in the occlusion pass — a pass that would otherwise just re-walk nodes `compute_dirty_regions` already walked. That is a lot of regression surface bought cheaply.

The PR's base (`45434a25b8`) also predates master's z-order work: master's `compute_dirty_regions` now carries `sibling_index` on every `CachedItemBoundingBoxAndTransform` variant plus `must_refresh_children |= sibling_index_changed` logic (master `partial_renderer.rs:453-529`), none of which exists in this rewrite. This will not rebase cleanly and the merged result has to be re-derived by hand. Please rebase and let us re-review that function specifically.

**8. `child_transform_to_screen` is 24 bytes per cached item to avoid one 2x3 matrix compose** (`partial_renderer.rs:182`). After `compute_dirty_regions`, `cached_geom.transform()` is already the current transform, so `compute_occlusion_recursive` can recompute `cached_geom.transform().then(&state.transform_to_screen)` for ~6 mul + 4 add. Trading per-item memory on a toolkit that targets MCUs for that is premature pessimization; please drop the field.

**9. No benchmark numbers for a feature whose sole purpose is performance.** The PR adds a full extra tree walk per frame plus per-item bookkeeping, and the Cargo.toml doc already hedges ("a straight cost on tight per-frame budgets"). Note also that marking a *container* occluded does not skip its children — `render_item_children` returns `ContinueRenderingChildren` when `do_draw` is false — so the win is limited to leaf draw calls while the walk cost is unconditional. Please post before/after frame times for a realistic occluded scene.

**10. Most of the per-crate `occlusion-culling` features are pass-through noise.** `i-slint-backend-winit`, `i-slint-backend-linuxkms` and `i-slint-backend-selector` each define an `occlusion-culling` feature that resolves to `i-slint-core/occlusion-culling`; Cargo feature unification already enables that everywhere the moment `slint`'s feature turns it on, so those three definitions change nothing. Only skia's and software's are load-bearing (they `#![cfg]` their test files). Collapse the `slint` feature to `occlusion-culling = ["i-slint-core/occlusion-culling", ...renderers-for-tests]`. Separately: there is no C++ CMake option and no docs-site entry, so the feature is Rust-only and undiscoverable.

## Minor

- `partial_renderer.rs:527` — "Only `Clip` items produce `ClipItem` (see `clips_children`), so this always downcasts" is not true: `Flickable::clips_children()` returns `true` (`internal/core/items/flickable.rs:236`), as do several text items. `unwrap_or_default()` keeps it correct, but the comment claims an invariant that doesn't hold. Say instead why a 0 inset is the right fallback for non-`Clip` clippers.
- `partial_renderer.rs:459` — "narrowed to the largest rectangle inscribed in every ancestor `Clip`'s rounded shape": inset-by-max-radius is *an* inscribed rectangle, not the largest one. The inline comment at 530 correctly says "approximation"; make the doc match.
- `item_rendering.rs:208` — "the number of items `render_item_children` has dispatched a draw call for" is inaccurate: the counter at line 271 also increments for `Clip`/`Opacity`/`Layer`/`BoxShadow`/`Transform` items entered with `do_draw == false`. "items whose `render()` was invoked" is accurate, and is also why the test thresholds are fuzzy.
- `item_rendering.rs:220,222,224` — `border_width <= LogicalLength::default()` / `border_radius() <= LogicalLength::default()`: `LogicalLength::zero()` is the clearer and more common spelling in this tree.
- `is_opaque_covering_rectangle` lives in `item_rendering.rs` but is `pub(crate)` with exactly one caller, in `partial_renderer.rs`. Move it next to its caller.
- `partial_renderer.rs:244` — `scan_and_collapse(&mut [Box2D], &mut usize, &Box2D)` passing a slice plus an out-of-band length is an awkward signature. Both callers are `{[T; N], usize}` pairs; a small inline `RectSet<const N: usize>` with `add`/`contains` would let `DirtyRegion` and `OccludedRegion` differ only in the overflow policy.
- `partial_renderer.rs:413` — `contains_rect` returns `true` for an empty rect, but `visible_rect` comes from `euclid::Rect::intersection`, which never yields an empty rect. Dead branch.
- `partial_renderer.rs:670,895` — with the feature off, `union_opt_rect` and the `aggregate` accumulation still run on `None` values throughout the mandatory dirty-region walk. The `screen_rect_for` doc claims "zero extra cost"; it is near-zero, not zero. Either `#[cfg]` the aggregate too, or soften the claim.
- `partial_renderer.rs:1058` — `drop(cache)` is an unrelated, unexplained change in `filter_item`. If it guards against a re-borrow, say so; otherwise drop it from this PR.
- `cspell.json:67` — `"AABB"` is inserted after `"aarch"`; the list is otherwise case-insensitively sorted, so it belongs before it.
- Both test files: `#[derive(Default)]` on the unit `TestPlatform` (skia) is unused, and `current_tick().0` is cast `as u64` in the software test but not in the skia one — the field is already `u64`, so the cast is redundant in both.
- `internal/renderers/software/tests/occlusion_culling.rs` — `NullLineBuffer::process_line` allocates a fresh `Vec` per line (200 per frame) and ignores `_line`. Hoist one scratch buffer into the struct.

## What's good

- The core algorithm is right and the reasoning is careful. Postorder over `FrontToBack` siblings is a correct reverse-paint-order walk, and the doc comment explaining *why* postorder (a parent paints before its children) saves the next reader real time.
- Region-based accumulation rather than sibling-only comparison is the right design, and the "different branch" test justifies it.
- The three soundness carve-outs — never treating a container's own bounds as coverage for unclipped overflowing children, insetting rounded `Clip`s by their corner radius, and clearing `may_contribute` under non-opaque `Opacity` and under `ItemWithTransform` — are exactly the three things a naive implementation gets wrong, and each has a named test.
- `OccludedRegion::add_rect` dropping rather than merging when full, with the doc comment spelling out that merging two disjoint rects would falsely claim the gap, is the correct asymmetry against `DirtyRegion` and is correctly explained.
- Factoring the shared containment/collapse scan out of `DirtyRegion::add_rect` instead of copy-pasting it was the right call.
- Comment quality throughout is above this repo's average: rationale rather than mechanics, with no redundant restating of the code.

**Comment by dfaure-kdab** [comment]
Both criticals are genuinely closed, and 16 of 22 previous findings are fully fixed — several better than suggested (RectSet<N> extraction, shared scenarios.rs).

## New findings

-  N1 partial_renderer.rs:644-647 — the device-pixel shrink over-claims under slint_int_coord. The final .cast() back to Rect<Coord, LogicalPx> truncates toward zero when Coord = i32, moving the min edge
  outward — the opposite of what round_in() was for. At scale 1.5, shrunk physical min 11 → logical 7.333 → 7; [7, 7.33) is claimed as covered but only partly painted. MCU builds only; the f32 path is
  correct (I verified the occludee side stays un-rounded). Do the shrink in Box2D with ceil()/floor(), or keep the accumulator in f32 logical coords.

-  N2 :624-646 — entry.occluded is still stale when visible_rect is None. The cache write sits inside the if let Some(visible_rect), so a fully-clipped item keeps last frame's flag. Same class as the
  fixed critical, reached via the clip path. One line: compute occluded with is_some_and and write it unconditionally, keeping the if let only for occluder recording.

-  N3 No test covers the anti-aliasing fix — both harnesses run at scale factor 1, so round_in() is the identity in every test. Add a set_scale_factor(1.3) scenario; without it N1 ships undetected.

-  N4 The two new test crates enable occlusion-culling workspace-wide. They dev-depend on i-slint-core with that feature; tests/Cargo.toml is resolver 3 and CI runs --all-features --workspace on it, so
  test-driver-rust, test-driver-interpreter and screenshots all now run with culling on. Arguably great coverage — but undeclared, so a culling bug will surface as an unrelated screenshot diff. Say so
  explicitly, or scope the feature to the two crates.

-  N5 The occlusion pass reads properties inside the window's redraw_tracker, unlike CachedItemBoundingBoxAndTransform::new and filter_item, which both use evaluate_no_tracking deliberately.
  is_opaque_covering_rectangle reads background/border_*, the walk reads Clip::logical_border_radius and Opacity::opacity. So bindings get evaluated for never-drawn items, and anything they transitively
  dirty misses this frame's region (the pass runs after compute_dirty_regions sampled is_dirty()). Wrap it, and add a step 1b to the module doc — the documented pipeline doesn't mention this pass at
  all.

-  Minor: N6 scenarios.rs:215 hardcodes + 4 for clip + border-radius lowering, which master is actively changing (1bb77eb3fc) — render each scene twice (opaque vs transparent occluder) and assert the
  difference. N7 stray blank-line deletion in selector/Cargo.toml:59. N8 MAX_COUNT: usize = 3 duplicates RectSet<3>. N9 empty [dependencies] in both new manifests.

##  Still open from before

-  7 partial — rebase is faithful (I diffed sibling_counters, the u16::MAX rule, same_geometry, old_transform, moved against master line by line), but the rewrite stands and the branch is 15 commits
  behind again. #9 caveat — the benchmark table may predate faa93c8324: fixing #1 moved the pass behind the region union, and on Skia back_buffer_age == 0 makes that the whole window, killing the prune.
  Ask which commit it was measured on. #10 partial — CMake option and C++ CHANGELOG added, still no docs-site entry. One 🟢 not addressed: union_opt_rect/aggregate still run with the feature off (the misleading "zero cost" claim was at least removed).

**Comment by dfaure-kdab** [comment]
All 12 open items fixed or disclosed.

  New findings

-  M1 partial_renderer.rs:633-637 — visible_rect hand-rolls transform+clip and skips the NaN guard, which the new int_coord code turns into a panic. Every other transform+clip here goes through transformed_and_clipped, which opens with if !rect.origin.is_finite() { return None }. Box2D::intersection returns Some for NaN (comparisons are false, so is_empty() is false), so a NaN occluder reaches the shrink — and the slint_int_coord arm's Box2D::cast::<i32>() is try_cast().unwrap(). Panic. Benign on the f32 build. The fix is a deletion: call transformed_and_clipped(cached_geom.bounding_rect(), state.transform_to_screen, &state.clipped) — behaviorally identical, its extra is_empty() only turns Some(empty) into None.

-  M2 The #[cfg(slint_int_coord)] arm written to fix N1 has never met a compiler. ci.yaml:394,399,405 are the only places setting that RUSTFLAG, and all three are cargo check --no-default-features on MCU demos — no occlusion-culling, so the whole pass is cfg'd out. The only thing enabling the feature is the tests/ workspace, which never sets the flag. Empty intersection. The branch reads correct (round_in() is ceil/floor in physical space; the extra logical ceil/floor can only shrink further, and skipping it on f32 is right — applying it there would coarsen to whole logical units), but that's not a type-check. Adding this to the CI config would fix it:
  ```
  - run: cargo check -p i-slint-core --features occlusion-culling
    env: { RUSTFLAGS: "--cfg slint_int_coord -D warnings" }
  ```

- Minor: M3 the fractional-scale test pins direction but not magnitude (over-shrinking still passes) — worth a doc line. M4 :659-676 needless Box2D → Rect → Box2D round trip; give OccludedRegion an add_box like DirtyRegion has. M5 backends_and_renderers.mdx:76 eats a blank line it needn't touch.

**Comment by R-Cramer4** [comment]
I am not sure if this is the best spot to put it in the ci

**Comment by dfaure-kdab** [comment]
> I am not sure if this is the best spot to put it in the ci

Claude says yes, because "this is where slint_int_coord lives", not because this needs any specific toolchain.
And placing it before the (time-consuming) xtensa toolchain install was a good call.

Claude also says well done for the tricky rebase on top of "repaint on any change in the z order of siblings" which landed meanwhile, all good, approved.

However unlike me Claude isn't an approver in this project, so @tronical or @ogoffart will need to have a look ;-)

File: internal/core/partial_renderer.rs [hunk 1]
```diff
@@ -399,144 +510,393 @@ pub struct PartialRenderer<'a, T> {
...
+    }
+}
+
+/// Screen-space bounding rect of an item's own rendering plus everything painted by its descendants,
+/// used only to feed `PartialRenderer::compute_occlusion`; a zero-sized no-op when that feature is off,
+/// so the bookkeeping below costs nothing to compile in.
+#[cfg(feature = "occlusion-culling")]
+type SubtreeScreenBounds = Option<LogicalRect>;
+#[cfg(not(feature = "occlusion-culling"))]
+type SubtreeScreenBounds = ();
```
Comment (the-argus) [inline internal/core/partial_renderer.rs]: Not sure why this whole cfg thing is here? It seems `SubtreeScreenBounds` is only used in `compute_dirty_regions_recursive`, which is only used under the `occlusion-culling` feature.

If it is absolutely needed, I would prefer just putting a cfg directive everywhere it is used rather than doing this.

File: internal/core/partial_renderer.rs [hunk 2]
```diff
@@ -399,144 +510,393 @@ pub struct PartialRenderer<'a, T> {
...
+// `SubtreeScreenBounds` collapses to `()` without the `occlusion-culling` feature, so this
+// bookkeeping is free in that build; that's also why it trips `let_unit_value` there.
+#[cfg_attr(not(feature = "occlusion-culling"), allow(clippy::let_unit_value))]
+fn compute_dirty_regions_recursive<T: ItemRendererFeatures>(
+    component: &ItemTreeRc,
+    index: isize,
+    cache: &RefCell<PartialRendererCache>,
+    window_adapter: &Rc<dyn WindowAdapter>,
+    dirty_region: &mut DirtyRegion,
+    // Two counters per tree depth to give each item its rank among its z-ordered siblings.
```
Comment (the-argus) [inline internal/core/partial_renderer.rs]: a long comment to explain tuple members probably indicates a struct with meaningful member names should be used

File: internal/core/partial_renderer.rs [hunk 3]
```diff
@@ -399,144 +510,393 @@ pub struct PartialRenderer<'a, T> {
...
+
+#[cfg(not(feature = "occlusion-culling"))]
+fn union_opt_rect(_a: SubtreeScreenBounds, _b: SubtreeScreenBounds) -> SubtreeScreenBounds {}
+
+/// Depth-first walk that computes dirty regions for `index`'s children
+/// Returns the union of `index`'s children's subtree bounds, or `None` if nothing under `index` is visible this frame.
+// `SubtreeScreenBounds` collapses to `()` without the `occlusion-culling` feature, so this
+// bookkeeping is free in that build; that's also why it trips `let_unit_value` there.
+#[cfg_attr(not(feature = "occlusion-culling"), allow(clippy::let_unit_value))]
+fn compute_dirty_regions_recursive<T: ItemRendererFeatures>(
```
Comment (the-argus) [inline internal/core/partial_renderer.rs]: this function gets heavily nested and has to have this conditionally unit return type thing, which I find confusing. some potential suggestions:

- break out the const parameters into some struct, break out bits of this such as the child visitor or heavily nested conditionals into seperate member functions
- avoid the weird conditional return type by not returning it and just having the callee do the rect union operations by reference

could also be done to compute_occlusion*, for consistency

File: tests/issues/test-1051-occlusion-culling/skia/tests/occlusion_culling.rs [hunk 4]
```diff
@@ -0,0 +1,145 @@
+// Copyright © Klarälvdalens Datakonsult AB, a KDAB Group company , info@kdab.com, author Robin Cramer <robin.cramer@kdab.com>
```
Comment (the-argus) [inline tests/issues/test-1051-occlusion-culling/skia/tests/occlusion_culling.rs:1]: Quite a few tests but IMO there should be more, they are cheap to write anyways, right?

- test for the usual/unoptimized case where occluder covers part of an item. assert that both items draw
- test occlusion when elements are not on top of each other due to their natural position but rather are forced via the z index property. could have a test for both that one with z index on top *does* occlude and z index below does not.
- test for any transparency effects and rounded corners (on the occluder). not sure what all transparency stuff slint has, but i imagine there is bg color alpha and maybe some gradient property which could also have alpha?
- test that an item which covers an other item normally, but it is rotated by 45 degrees so the corners of the item underneath are visible
- test that occlusion stops when the occluder moves after the first draw / occlusion calculation, and that occlusion happens if the occluder moves over top of something on a frame after the first one is drawn.

File: internal/core/partial_renderer.rs [hunk 5]
```diff
@@ -399,144 +510,393 @@ pub struct PartialRenderer<'a, T> {
...
+
+        // Computed whether or not `visible_rect` is `Some`, and always written back below: a
+        // fully-clipped item (`visible_rect` is `None`) must not keep last frame's flag.
+        let occluded =
+            visible_rect.is_some_and(|visible_rect| accumulator.contains_rect(visible_rect));
+
+        {
+            let mut cache = cache.borrow_mut();
+            if let Some(entry) = rendering_data.get_entry(&mut cache) {
+                entry.occluded = occluded;
```
Comment (the-argus) [inline internal/core/partial_renderer.rs:691]: I think in this case there needs to be something like `if (occluded) { entry.tracker = None; }`, otherwise dirty + occluded items will continue to be dirty, provided they were drawn before and then covered up

**Review by the-argus — COMMENTED** [review]
Seems good, here are some slightly pedantic things, plus a minor bug fable noticed :)

**Comment by the-argus** [comment]
Thanks for the fixes, I found it much easier to read the occlusion pass after the changes :D 

Breaking up the dirty region pass seems a worthwhile improvement as well IMO, though I suppose it is a little bit of an increase in the scope of this to refactor that.

So LGTM too, we will see what others think after the 1.18 release
